utils/urlout.1
.\" urlout - stream output of url
.\" Copyright (C) 2013-2014 David Sugar <dyfet@gnutelephony.org>
.\" Copyright (C) 2015 Cherokees of Idaho.
.\"
.\" This manual page is free software; you can redistribute it and/or modify
.\" it under the terms of the GNU General Public License as published by
.\" the Free Software Foundation; either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" This program is distributed in the hope that it will be useful,
.\" but WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
.\" GNU General Public License for more details.
.\"
.\" You should have received a copy of the GNU Lesser General Public License
.\" along with this program.  If not, see <http://www.gnu.org/licenses/>.
.\"
.\" This manual page is written especially for Debian GNU/Linux.
.\"
.TH urlout "1" "August 2013" "GNU uCommon" "GNU Telephony"
.SH NAME
urlout \- stream output of url.
.SH SYNOPSIS
.B urlout
.RI [ options ]
.RI url
.br
.SH DESCRIPTION
This utility will stream a url address to stdout.  It can be used to test
usecure ssl/tls streaming operations, as well as supporting generic http
urls.  Various options will be added to aide debugging of tls streams in
the future.
.SH OPTIONS
.TP
.BI \-\-port= numnber
Specify alternate port number.
.TP
.B \-\-help
Outputs help screen for the user.
.SH AUTHOR
.B urlout
was written by David Sugar <dyfet@gnutelephony.org>.
.SH "REPORTING BUGS"
Report bugs to bug-commoncpp@gnu.org or bugs@gnutelephony.org.
.SH COPYRIGHT
Copyright \(co 2013-2014 David Sugar, Tycho Softworks.
.br
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR
PURPOSE.



ChangeLog
Changes from 6.6.0 to 7.0.0
- clean new library abi 8.0
- atomic typeref system completed
- c++11 support completed
- deprecated functions and templates removed
- deprecated modules (xml, persist) moved to commoncpp

Changes from 6.5.7 to 6.6.0
- introduced rsa key support
- expanded hmac support
- expanded digests for sha256 and 384
- reword of common digest code
- improved nullptr clang support
- remove clang forced c++11 from build
- check for openssl rsa support
- port types for 7.0 migration
- socket addresses for typeref
- further c++11 header fixes
- removed old ssl demo app

Changes from 6.5.6 to 6.5.7
- improved c++11 support
- mapped pointer introduced
- fixed is usage
- improved mapref remove

Changes from 6.5.5 to 6.5.6
- simplified arrayref
- added listref
- map iterators thru locked instances
- type standardization
- socket address type

Changes from 6.5.4 to 6.5.5
- more portable nullptr support
- thread-safe mapref class 
- some typeref convenience types
- bit operations on byterefs

Changes from 6.5.3 to 6.5.4
- secure string and key management types
- better cleanup of secure objects
- file i/o for heap temporary

Changes from 6.5.2 to 6.5.3
- arrayref now uses ConditionalAccess, fix for Conditional

Changes from 6.5.1 to 6.5.2
- memory management cleanup and mingw32 support for native conditionals
- new methodology of having getaddrinfo allocate memory
- introduction of queueref and stackref; arrayref becomes useful
- typeref concatenation operators

Changes from 6.5.0 to 6.5.1
- thread shared references added

Changes from 6.4.4 to 6.5.0
- typeref expanded
- arrayref introduced
- nullptr and other c++ modernizations
- clang now defaulted to c++11
- minimum native windows now requires conditionals
- mingw has to use win32 pthread support
- somewhat more usable heap temporary templates

Changes from 6.4.3 to 6.4.4
- additional typeref operators

Changes from 6.4.2 to 6.4.3
- fix for broken windows setuid macro

Changes from 6.4.1 to 6.4.2
- solaris related cmake fixes
- cleanup of test build and osx fixes

Changes from 6.4.0 to 6.4.1
- keyfile fixed constructor issue

Changes from 6.3.6 to 6.4.0
- new typeref system for immutable atomic reference counted objects
- heap management objects support moving heap through assignment
- extended unit tests for typeref system
- improved openbsd support
- atomics enabled by default

Changes from 6.3.5 to 6.3.6
- code cleanup
- simulate option for scrub
- set newline style for cmake genorated files

Changes from 6.3.4 to 6.3.5
- general code cleanup
- some build fixes

Changes from 6.3.3 to 6.3.4
- improved atomics support
- configure atomics default matches cmake default
- general code cleanup
- more casting operations and cast fixups
- polymorphic casting support & rtti detection
- enclose random value templates in Random
- improved rng support

Changes from 6.3.2 to 6.3.3
- improved cipher key management
- b64 support improved and string hex conversions
- simplified digest functions
- some solaris fixes

Changes from 6.3.1 to 6.3.2
- fixed a broken streambuf for commoncpp
- deref cast function added
- fixed missing pkg-config

Changes from 6.3.0 to 6.3.1
- copyright assignment to Cherokees of Idaho
- introduction of ucommon cmake hosted macros
- simplified doxygen doc support
- convergence of automake and cmake builds
- make utils easier for source introspection
- converted for cmake builds on debian
- count for String::check should be size_t
- string bounds checking

Changes from 6.2.3 to 6.3.0
- remove legacy cmodel support
- modernize library usage
- simplify autoconf and cmake
- introduce sysruntime...
- cleanup memory operators....

Changes from 6.2.2 to 6.2.3
- a few socket fixes, mostly windows related...

Changes from 6.2.1 to 6.2.2
- bumped gnutls to 3.0.0 or later...
- lots of bug fixes and general cleanup
- socket endian issues and other things for ccrtp
- templated newp and freep handle management

Changes from 6.2.0 to 6.2.1
- datetime: fix const inconsistency
- fsys: cached error variable must be mutable
- replacment new/delete cannot be declared inline
- fix exception specifier clash for c++11

Changes from 6.1.11 to 6.2.0
- standardized cmake paths and stand-alone usages
- gnutls support in cmake
- unified openssl checks, even works for windows
- improved openssl support, new static crypto option
- pkgconfig usable in win32, other win32 fixes
- fixes for mac build & optional disable of utils
- upticked abi version to 7, should have happened with 10/11?

Changes from 6.1.10 to 6.1.11
- any addr and overloads in commoncpp
- cmake abi fix and other apple specific issues
- stream fixes, sflphone recommended changes, and other updates
- key protocol added
- cleaner build for binding without resolver
- cleanup and fixes for commoncpp peering...
- UDPSocket: use ucommon::Socket::address
- Socket::address: add setAny(sockaddr*), setLoopback(sockaddr*)

Changes from 6.1.9 to 6.1.10
- for next release
- iso cleaness
- Fix memory leak in DateTime ctor
- Fix utf8::count()
- secure::erase() should be more secure

Changes from 6.1.8 to 6.1.9
- ost::Socket: use ucommon::Socket for join(), drop() implementations
- back to native threading model for mingw
- Socket::store: use existing len method
- Socket::address: allow printing to std::ostream
- Socket::address: use in_port_t for port numbers
- Socket::join, ::drop: allow to set interface index for IPv6

Changes from 6.1.7 to 6.1.8
- mingw libstdc++ c++11 support now requires winpthread
- more android related updates

Changes from 6.1.6 to 6.1.7
- lots of android related updates

Changes from 6.1.5 to 6.1.6
- some windows and mingw fixes
- Socket::address::print(): minor fixes
- Socket::address: fix constructor documentation
- Socket::address: add unit tests for print()
- Socket::address: add print() method
- Socket::address: add withPort()
- Socket::address: add constructors from standard addr types
- Socket::address: use existing len() method.

Changes from 6.1.4 to 6.1.5

utils/urlout.cpp
#include <ucommon/secure.h>
#include <sys/stat.h>

using namespace ucommon;

static shell::flagopt helpflag('h',"--help",    _TEXT("display this list"));
static shell::flagopt althelp('?', NULL, NULL);
static shell::numericopt port('p', "--port", _TEXT("port to use"), "port", 0);

int main(int argc, char **argv)
{
    shell::bind("urlout");
    shell args(argc, argv);

    if(is(helpflag) || is(althelp) || !args() || args() > 1) {
        printf("%s\n", _TEXT("Usage: urlout [options] url-path"));
        printf("\n%s\n", _TEXT("Options:"));
        shell::help();
        printf("\n%s\n", _TEXT("Report bugs to dyfet@gnu.org"));
        return 0;
    }

    bool url_secure = false;
    const char *url = args[0];
    const char *proto = "80";
    secure::client_t ctx = NULL;
    String svc;
    
    if(eq(url, "https://", 8)) {
        url_secure = true;
        url += 8;
    }
    else if(eq(url, "http://", 7)) {
        url += 7;
    }
    else
        url_secure = true;

    string_t host = url;
    string_t path = strdup(url);

    if(url_secure && secure::init()) {
        proto = "443";
        ctx = secure::client();
    }

    if(is(port))
        svc = str(*port);
    else
        svc = proto;

    const char *find = path.find("/");
    if(find)
        path.rsplit(find);
    else
        path = "/";

    host.split(host.find("/"));

    sstream web(ctx);
    web.open(host, svc);

    web << "GET /\r\n\r\n";
    web.flush();

    std::cout << web.rdbuf();
}
